File: 动态规划/最大子序和.py
# coding: utf-8

class Solution(object):
    def maxSubArray(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if len(nums) == 0:
            return 0

        res = nums[0]
        F_Max = nums[0]
        for i in range(1, len(nums)):
            F_Max = max(F_Max + nums[i], nums[i])
            res = max(res, F_Max)

        return res

nums = [-2,1,-3,4,-1,2,1,-5,4]
print(Solution().maxSubArray(nums))

File: old_bak/golang/junior/最大子序和.go
package main

func maxSubArray(nums []int) int {
	if len(nums) == 0 {
		return 0
	}

	res := nums[0]
	var max = res
	for i := 1; i < len(nums); i++ {
		tmp := res + nums[i]
		if tmp > nums[i] {
			res = tmp
		} else {
			res = nums[i]
		}

		if res > max {
			max = res
		}
	}

	return max
}
